CMakeLists.txt

```diff
@@ -1,367 +1,165 @@
-
-
-###
-### General install settings
-###
-set(INCLUDE_INSTALL_ROOT_DIR include)
-
-set(INCLUDE_INSTALL_DIR ${INCLUDE_INSTALL_ROOT_DIR}/yaml-cpp)
-set(LIB_INSTALL_DIR "lib${LIB_SUFFIX}")
-
-set(_INSTALL_DESTINATIONS
-	RUNTIME DESTINATION bin
-	LIBRARY DESTINATION ${LIB_INSTALL_DIR}
-	ARCHIVE DESTINATION ${LIB_INSTALL_DIR}
-)
-
-
-###
-### Library
-###
-if(YAML_BUILD_SHARED_LIBS)
-    add_library(yaml-cpp SHARED ${library_sources})
-else()
-    add_library(yaml-cpp STATIC ${library_sources})
-endif()
-
-if (NOT CMAKE_VERSION VERSION_LESS 2.8.12)
-    target_include_directories(yaml-cpp
-        PUBLIC $<BUILD_INTERFACE:${YAML_CPP_SOURCE_DIR}/include>
-               $<INSTALL_INTERFACE:${INCLUDE_INSTALL_ROOT_DIR}>
-        PRIVATE $<BUILD_INTERFACE:${YAML_CPP_SOURCE_DIR}/src>)
-endif()
-
-set_target_properties(yaml-cpp PROPERTIES
-    CXX_STANDARD 11
-    CXX_STANDARD_REQUIRED ON
-)
-
-set_target_properties(yaml-cpp PROPERTIES
-  COMPILE_FLAGS "${yaml_c_flags} ${yaml_cxx_flags}"
-)
+cmake_dependent_option(YAML_CPP_BUILD_TESTS
+  "Enable yaml-cpp tests" ON
+  "BUILD_TESTING;CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR" OFF)
+cmake_dependent_option(YAML_CPP_INSTALL
+  "Enable generation of yaml-cpp install targets" ON
+  "CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR" OFF)
+cmake_dependent_option(YAML_BUILD_SHARED_LIBS
+  "Build yaml-cpp shared library" OFF
+  "BUILD_SHARED_LIBS" OFF)
+
+cmake_dependent_option(YAML_MSVC_SHARED_RT
+  "MSVC: Build yaml-cpp with shared runtime libs (/MD)" ON
+  "MSVC" OFF)
+cmake_dependent_option(YAML_APPLE_UNIVERSAL_BIN
+  "Apple: Build yaml-cpp universal binary" OFF
+  "APPLE" OFF)
+
+set(yaml-cpp-type STATIC)
+if (YAML_BUILD_SHARED_LIBS)
+  set(yaml-cpp-type SHARED)
+endif()
+
+set(build-shared $<BOOL:${YAML_BUILD_SHARED_LIBS}>)
+set(build-windows-dll $<AND:$<BOOL:${CMAKE_HOST_WIN32}>,${build-shared}>)
+set(not-msvc $<NOT:$<CXX_COMPILER_ID:MSVC>>)
+
+if (NOT DEFINED CMAKE_MSVC_RUNTIME_LIBRARY)
+  set(CMAKE_MSVC_RUNTIME_LIBRARY
+    MultiThreaded$<$<CONFIG:Debug>:Debug>$<${build-shared}:DLL>)
+endif()
+
+set(contrib-pattern "src/contrib/*.cpp")
+set(src-pattern "src/*.cpp")
+if (CMAKE_VERSION VERSION_GREATER 3.12)
+  list(INSERT contrib-pattern 0 CONFIGURE_DEPENDS)
+  list(INSERT src-pattern 0 CONFIGURE_DEPENDS)
+endif()
+
+file(GLOB yaml-cpp-contrib-sources ${contrib-pattern})
+file(GLOB yaml-cpp-sources ${src-pattern})
+
+# /wd4127 = disable warning C4127 "conditional expression is constant"
+# http://msdn.microsoft.com/en-us/library/6t66728h.aspx
+# /wd4355 = disable warning C4355 "'this' : used in base member initializer list
+# http://msdn.microsoft.com/en-us/library/3c594ae3.aspx
+set(msvc-rt $<TARGET_PROPERTY:MSVC_RUNTIME_LIBRARY>)
+
+set(mtd-static $<STREQUAL:${msvc-rt},MultiThreadedDebug>)
+set(mt-static $<STREQUAL:${msvc-rt},MultiThreaded>)
+
+set(mtd-dll $<STREQUAL:${msvc-rt},MultiThreadedDebugDLL>)
+set(mt-dll $<STREQUAL:${msvc-rt},MultiThreadedDLL>)
+
+set(backport $<VERSION_LESS:${CMAKE_VERSION},3.15>)
```

Comment:
Same here.

Reply:
This is being done because CMake 3.15 adds support for setting `CMAKE_MSVC_RUNTIME_LIBRARY`. We don't want to step on their toes for that, thus we only enable this setting if someone is using pre-3.15. Otherwise we could have `target_compile_options` and cmake generation conflict and cause link time errors. All the logic regarding the `backport` variable is just done to make sure that it gets eliminated early on newer CMake versions.

Reply:
Gotcha. Can you rename this variable, something like add_msvc_runtime_option (or whatever, something that says what it does)?

Reply:
Sure! I'll update the names in a bit and do a push when I have a moment later today.

Reply:
Names have been updated! Let me know if it meets your needs :)